File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []
        def backtrack(start , li , sum ,  target ):
            nonlocal ans , candidates 
            if sum == target:
                ans.append(li)
                return 
            elif sum > target:
                return 
            for i in range(start, len(candidates)):
                    backtrack(i ,li+[candidates[i]] , sum + candidates[i] , target)
        backtrack(0,[] , 0 , target)
        return ans
            

File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    vector<vector<int>> ans;
    vector<int> li;
    void traverse(int start  , vector<int>& candidates , int su){
        if (su<0) return ;
        if (su == 0 ){
            ans.push_back(li);
            return ;
        }
        for (int i = start ; i<candidates.size() ; i++){
            li.push_back(candidates[i]);
            traverse(i, candidates , su - candidates[i] );
            li.pop_back();
        }
        return ;
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target){
        traverse(0 , candidates , target);
        return ans;
    }
};